compiler/noirc_frontend/src/hir/comptime/interpreter.rs

```diff
@@ -0,0 +1,1282 @@
+                        Err(InterpreterError::NonPointerDereferenced { value, location: *location })
+                    }
+                }
+            }
+            HirLValue::MemberAccess { object, field_name, field_index, typ: _, location } => {
+                let index = field_index.expect("The field index should be set after type checking");
+
+                match self.evaluate_lvalue(object)? {
+                    Value::Tuple(mut values) => Ok(values.swap_remove(index)),
+                    Value::Struct(fields, _) => Ok(fields[&field_name.0.contents].clone()),
+                    value => Err(InterpreterError::NonTupleOrStructInMemberAccess {
+                        value,
+                        location: *location,
+                    }),
+                }
+            }
+            HirLValue::Index { array, index, typ: _, location } => {
+                let array = self.evaluate_lvalue(array)?;
+                let index = self.evaluate(*index)?;
+                let (elements, index) = self.bounds_check(array, index, *location)?;
+                Ok(elements[index].clone())
+            }
+        }
+    }
+
+    fn evaluate_for(&mut self, for_: HirForStatement) -> IResult<Value> {
+        // i128 can store all values from i8 - u64
+        let get_index = |this: &mut Self, expr| -> IResult<(_, fn(_) -> _)> {
+            match this.evaluate(expr)? {
+                Value::I8(value) => Ok((value as i128, |i| Value::I8(i as i8))),
+                Value::I32(value) => Ok((value as i128, |i| Value::I32(i as i32))),
+                Value::I64(value) => Ok((value as i128, |i| Value::I64(i as i64))),
+                Value::U8(value) => Ok((value as i128, |i| Value::U8(i as u8))),
+                Value::U32(value) => Ok((value as i128, |i| Value::U32(i as u32))),
+                Value::U64(value) => Ok((value as i128, |i| Value::U64(i as u64))),
+                value => {
+                    let location = this.interner.expr_location(&expr);
+                    Err(InterpreterError::NonIntegerUsedInLoop { value, location })
+                }
+            }
+        };
+
+        let (start, make_value) = get_index(self, for_.start_range)?;
+        let (end, _) = get_index(self, for_.end_range)?;
+        let was_in_loop = std::mem::replace(&mut self.in_loop, true);
+
+        for i in start..end {
+            self.push_scope();
+            self.current_scope_mut().insert(for_.identifier.id, make_value(i));
+
+            match self.evaluate(for_.block) {
+                Ok(_) => (),
+                Err(InterpreterError::Break) => break,
+                Err(InterpreterError::Continue) => continue,
+                Err(other) => return Err(other),
+            }
+            self.pop_scope();
+        }
+
+        self.in_loop = was_in_loop;
+        Ok(Value::Unit)
+    }
+
+    fn evaluate_break(&mut self) -> IResult<Value> {
+        if self.in_loop {
+            Err(InterpreterError::Break)
```

Comment:
Should we just have a separate enum for the loop state. It can be like:
```rust
enum LoopState {
     InProgress,
     Break,
     Continue
}
```
It feels weird having this continue state as an error and not simply propagating it. Then inside of `evaluate_for` we can just have this:
```rust
match self.evaluate(for_.block)? {
    LoopState::InProgress => {},
    LoopState::Break => break,
    LoopState::Continue => continue,
}
```

Reply:
We'd have to return a `(Value, LoopState)` from every function in that case which was why I initially avoided it

Reply:
Yeah I see. What if we make `LoopState` a `Value` though?

Reply:
> You'd also have to match on it in every function like an error value to ensure you don't continue evaluating if a break or continue was returned.

Because shouldn't this be handled by the `in_loop` flag then and we return an error?

Reply:
If `LoopState` is a Value it'd still have to be matched on and exited early from every function. Most functions match on values anyway so this would mostly add two
```
Value::Break => return Value::Break,
Value::Continue => return Value::Continue,
```
branches to every match case. Certain functions which ignore values (e.g. `StatementKind::Semi`) would need to have matches added to propagate the break/continue.

> Because shouldn't this be handled by the in_loop flag then and we return an error?
The `in_loop` flag is just for issuing `BreakNotInLoop` errors, you'd still need to match on `Value::Break` with or without it.

Reply:
ah yeah

Reply:
Ok we can leave it as an Error then